// circom/parser/jaz.js
/* parser generated by jison 0.4.18 */
/*
  Returns a Parser object of the following structure:

  Parser: {
    yy: {}
  }

  Parser.prototype: {
    yy: {},
    trace: function(),
    symbols_: {associative list: name ==> number},
    terminals_: {associative list: number ==> name},
    productions_: [...],
    performAction: function anonymous(yytext, yyleng, yylineno, yy, yystate, $$, _$),
    table: [...],
    defaultActions: {...},
    parseError: function(str, hash),
    parse: function(input),

    lexer: {
        EOF: 1,
        parseError: function(str, hash),
        setInput: function(input),
        input: function(),
        unput: function(str),
        more: function(),
        less: function(n),
        pastInput: function(),
        upcomingInput: function(),
        showPosition: function(),
        test_match: function(regex_match_array, rule_index),
        next: function(),
        lex: function(),
        begin: function(condition),
        popState: function(),
        _currentRules: function(),
        topState: function(),
        pushState: function(condition),

        options: {
            ranges: boolean           (optional: true ==> token location info will include a .range[] member)
            flex: boolean             (optional: true ==> flex-like lexing behaviour where the rules are tested exhaustively to find the longest match)
            backtrack_lexer: boolean  (optional: true ==> lexer regexes are tested in order and for each matching regex the action code is invoked; the lexer terminates the scan when a token is returned by the action code)
        },

        performAction: function(yy, yy_, $avoiding_name_collisions, YY_START),
        rules: [...],
        conditions: {associative list: name ==> set},
    }
  }


  token location info (@$, _$, etc.): {
    first_line: n,
    last_line: n,
    first_column: n,
    last_column: n,
    range: [start_number, end_number]       (where the numbers are indexes into the input string, regular zero-based)
  }


  the parseError function receives a 'hash' object with these members for lexer and parser errors: {
    text:        (matched text)
    token:       (the produced terminal token, if any)
    line:        (yylineno)
  }
  while parser (grammar) errors will also provide these members, i.e. parser errors deliver a superset of attributes: {
    loc:         (yylloc)
    expected:    (string describing the set of expected tokens)
    recoverable: (boolean: TRUE when the parser has a error recovery rule available for this particular error)
  }
*/
var jaz = (function(){
var o=function(k,v,o,l){for(o=o||{},l=k.length;l--;o[k[l]]=v);return o},$V0=[1,15],$V1=[1,33],$V2=[1,64],$V3=[1,16],$V4=[1,17],$V5=[1,18],$V6=[1,19],$V7=[1,20],$V8=[1,21],$V9=[1,25],$Va=[1,22],$Vb=[1,23],$Vc=[1,53],$Vd=[1,54],$Ve=[1,51],$Vf=[1,52],$Vg=[1,55],$Vh=[1,56],$Vi=[1,62],$Vj=[1,63],$Vk=[1,34],$Vl=[1,35],$Vm=[1,36],$Vn=[1,37],$Vo=[1,60],$Vp=[1,61],$Vq=[1,40],$Vr=[5,18,19,20,23,25,28,30,31,32,33,35,36,37,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115],$Vs=[5,18,19,20,23,25,27,28,30,31,32,33,35,36,37,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115],$Vt=[1,78],$Vu=[1,79],$Vv=[1,80],$Vw=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,79,81,82,84,85,87,88,89,90,94,95,99,100,104,105,106,110,112,113,115,116],$Vx=[2,97],$Vy=[1,94],$Vz=[1,95],$VA=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,57,58,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VB=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,79,81,82,84,85,87,88,89,90,92,93,94,95,99,100,104,105,106,110,112,113,116],$VC=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VD=[2,112],$VE=[1,102],$VF=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VG=[1,110],$VH=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VI=[1,113],$VJ=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VK=[1,114],$VL=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,67,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VM=[1,115],$VN=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,67,69,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VO=[1,116],$VP=[1,117],$VQ=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,67,69,71,73,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VR=[1,118],$VS=[1,119],$VT=[1,120],$VU=[1,121],$VV=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VW=[1,122],$VX=[1,123],$VY=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,79,81,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$VZ=[1,124],$V_=[1,125],$V$=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,79,81,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$V01=[1,126],$V11=[1,127],$V21=[1,128],$V31=[1,129],$V41=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,79,81,82,84,85,87,88,89,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$V51=[1,130],$V61=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,52,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,79,81,82,84,85,87,88,89,90,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$V71=[1,132],$V81=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,79,81,82,84,85,87,88,89,90,92,93,94,95,99,100,102,104,105,106,110,112,113,115,116],$V91=[1,179],$Va1=[22,24,116],$Vb1=[1,203],$Vc1=[5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,58,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],$Vd1=[1,219],$Ve1=[22,24];
var parser = {trace: function trace () { },
yy: {},
symbols_: {"error":2,"allStatments":3,"statmentList":4,"EOF":5,"statment":6,"functionDefinitionStatment":7,"templateDefinitionStatment":8,"ifStatment":9,"forStatment":10,"whileStatment":11,"doWhileStatment":12,"returnStatment":13,"block":14,"computationBlock":15,"expressionStatment":16,"includeStatment":17,"function":18,"IDENTIFIER":19,"(":20,"identifierList":21,")":22,"template":23,",":24,"if":25,"expression":26,"else":27,"for":28,";":29,"while":30,"do":31,"return":32,"include":33,"STRING":34,"{":35,"}":36,"compute":37,"e17":38,"leftHandExpression":39,"=":40,"+=":41,"-=":42,"*=":43,"/=":44,"%=":45,"<<=":46,">>=":47,"&=":48,"|=":49,"^=":50,"<==":51,"==>":52,"<--":53,"-->":54,"e16":55,"===":56,"?":57,":":58,"rightArray":59,"e15":60,"||":61,"e14":62,"&&":63,"e13":64,"|":65,"e12":66,"^":67,"e11":68,"&":69,"e10":70,"==":71,"e9":72,"!=":73,"<=":74,"e7":75,">=":76,"<":77,">":78,"<<":79,"e6":80,">>":81,"+":82,"e5":83,"-":84,"*":85,"e4":86,"/":87,"\\\\":88,"%":89,"**":90,"e3":91,"++":92,"--":93,"!":94,"~":95,"e2":96,"functionCall":97,"e0":98,"DECNUMBER":99,"HEXNUMBER":100,"simpleLeftHandExpression":101,".":102,"declaration":103,"var":104,"int":105,"signal":106,"input":107,"private":108,"output":109,"component":110,"array":111,"field":112,"log":113,"expressionList":114,"[":115,"]":116,"$accept":0,"$end":1},
terminals_: {2:"error",5:"EOF",18:"function",19:"IDENTIFIER",20:"(",22:")",23:"template",24:",",25:"if",27:"else",28:"for",29:";",30:"while",31:"do",32:"return",33:"include",34:"STRING",35:"{",36:"}",37:"compute",40:"=",41:"+=",42:"-=",43:"*=",44:"/=",45:"%=",46:"<<=",47:">>=",48:"&=",49:"|=",50:"^=",51:"<==",52:"==>",53:"<--",54:"-->",56:"===",57:"?",58:":",61:"||",63:"&&",65:"|",67:"^",69:"&",71:"==",73:"!=",74:"<=",76:">=",77:"<",78:">",79:"<<",81:">>",82:"+",84:"-",85:"*",87:"/",88:"\\\\",89:"%",90:"**",92:"++",93:"--",94:"!",95:"~",99:"DECNUMBER",100:"HEXNUMBER",102:".",104:"var",105:"int",106:"signal",107:"input",108:"private",109:"output",110:"component",112:"field",113:"log",115:"[",116:"]"},
productions_: [0,[3,2],[4,2],[4,1],[6,1],[6,1],[6,1],[6,1],[6,1],[6,1],[6,1],[6,1],[6,1],[6,1],[6,1],[7,6],[7,5],[8,6],[8,5],[21,3],[21,1],[9,7],[9,5],[10,9],[11,5],[12,6],[13,3],[13,2],[17,3],[17,2],[14,3],[15,4],[16,2],[16,1],[26,1],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,3],[38,5],[38,1],[55,1],[55,1],[60,3],[60,1],[62,3],[62,1],[64,3],[64,1],[66,3],[66,1],[68,3],[68,1],[70,3],[70,3],[70,1],[72,3],[72,3],[72,3],[72,3],[72,1],[75,3],[75,3],[75,1],[80,3],[80,3],[80,1],[83,3],[83,3],[83,3],[83,3],[83,1],[86,3],[86,1],[91,2],[91,2],[91,2],[91,2],[91,2],[91,2],[91,1],[96,2],[96,2],[96,1],[96,1],[98,1],[98,1],[98,1],[98,3],[39,3],[39,1],[39,1],[103,2],[103,2],[103,2],[103,3],[103,4],[103,3],[103,2],[101,2],[101,1],[97,4],[97,4],[97,4],[97,4],[97,3],[114,3],[114,1],[59,1],[111,3]],
performAction: function anonymous(yytext, yyleng, yylineno, yy, yystate /* action[1] */, $$ /* vstack */, _$ /* lstack */) {
/* this == yyval */

var $0 = $$.length - 1;
switch (yystate) {
case 1:

//            console.log(JSON.stringify($$[$0-1], null, 1));
            this.$ = { type: "BLOCK", statements: $$[$0-1].statments };
            setLines(this.$, _$[$0-1]);
            return this.$
        
break;
case 2:

            $$[$0-1].statments.push($$[$0]);
            setLines($$[$0-1], _$[$0-1], _$[$0]);
        
break;
case 3:

            this.$ = { type: "STATMENTLIST", statments: [$$[$0]] };
            setLines(this.$, _$[$0]);
        
break;
case 4: case 5: case 6: case 7: case 8: case 9: case 10: case 11: case 12: case 13: case 14: case 33: case 34: case 62: case 64: case 78: case 83: case 85: case 92: case 95: case 96: case 120:

            this.$ = $$[$0];
        
break;
case 15:

            this.$ = { type: "FUNCTIONDEF", name: $$[$0-4], params: $$[$0-2].identifiers, block: $$[$0]};
            setLines(this.$, _$[$0-5], _$[$0]);
        
break;
case 16:

            this.$ = { type: "FUNCTIONDEF", name: $$[$0-3], params: [], block: $$[$0] };
            setLines(this.$, _$[$0-4], _$[$0]);
        
break;
case 17:

            this.$ = { type: "TEMPLATEDEF", name: $$[$0-4], params: $$[$0-2].identifiers, block: $$[$0] };
            setLines(this.$, _$[$0-5], _$[$0]);
        
break;
case 18:

            this.$ = { type: "TEMPLATEDEF", name: $$[$0-3], params: [], block: $$[$0] };
            setLines(this.$, _$[$0-4], _$[$0]);
        
break;
case 19:

            $$[$0-2].identifiers.push($$[$0]);
            setLines($$[$0-2], _$[$0-2], _$[$0]);
        
break;
case 20:

            this.$ = { type: "IDENTIFIERLIST", identifiers: [$$[$0]] };
            setLines(this.$, _$[$0]);
        
break;
case 21:

            if ($$[$0-4].type == "NUMBER") {
                this.$ = !$$[$0-4].value.eq(0) ? $$[$0-2] : $$[$0];
            } else {
                this.$ = { type: "IF", condition: $$[$0-4], then: $$[$0-2], else: $$[$0] };
            }
            setLines(this.$, _$[$0-6], _$[$0]);
        
break;
case 22:

            if ($$[$0-2].type == "NUMBER") {
                this.$ = !$$[$0-2].value.eq(0) ? $$[$0] : { type: "NUMBER", value: bigInt(0) };
            } else {
                this.$ = { type: "IF", condition: $$[$0-2], then: $$[$0] };
            }
            setLines(this.$, _$[$0-4], _$[$0]);
        
break;
case 23:

            this.$ = { type: "FOR", init: $$[$0-6], condition: $$[$0-4], step: $$[$0-2], body: $$[$0] };
            setLines(this.$, _$[$0-8], _$[$0]);
        
break;
case 24:

            this.$ = { type: "WHILE", condition: $$[$0-2], body: $$[$0] };
            setLines(this.$, _$[$0-4], _$[$0]);
        
break;
case 25:

            this.$ = { type: "DOWHILE", condition: $$[$0-1], body: $$[$0-4] };
            setLines(this.$, _$[$0-5], _$[$0]);
        
break;
case 26:

            this.$ = { type: "RETURN", value: $$[$0-1] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 27:

            this.$ = { type: "RETURN", value: $$[$0] }
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 28:

            this.$ = { type: "INCLUDE", file: $$[$0-1] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 29:

            this.$ = { type: "INCLUDE", file: $$[$0] }
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 30:

            this.$ = { type: "BLOCK", statements: $$[$0-1].statments };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 31:

            this.$ = { type: "COMPUTEBLOCK", statements: $$[$0-1].statments };
            setLines(this.$, _$[$0-3], _$[$0]);
        
break;
case 32:

            this.$ = $$[$0-1];
        
break;
case 35:

            this.$ = { type: "OP", op: "=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 36:

            this.$ = { type: "OP", op: "+=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 37:

            this.$ = { type: "OP", op: "-=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 38:

            this.$ = { type: "OP", op: "*=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 39:

            this.$ = { type: "OP", op: "/=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 40:

            this.$ = { type: "OP", op: "%=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 41:

            this.$ = { type: "OP", op: "<<=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 42:

            this.$ = { type: "OP", op: ">>=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 43:

            this.$ = { type: "OP", op: "&=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 44:

            this.$ = { type: "OP", op: "|=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 45:

            this.$ = { type: "OP", op: "^=", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 46:

            this.$ = { type: "OP", op: "<==", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 47:

            this.$ = { type: "OP", op: "<==", values: [$$[$0], $$[$0-2]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 48:

            this.$ = { type: "OP", op: "<--", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 49:

            this.$ = { type: "OP", op: "<--", values: [$$[$0], $$[$0-2]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 50:

            this.$ = { type: "OP", op: "===", values: [$$[$0-2], $$[$0]] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 51:

            if ($$[$0-4].type == "NUMBER") {
                this.$ = !$$[$0-4].value.eq(0) ? $$[$0-2] : $$[$0];
            } else {
                this.$ = { type: "OP", op: "?", values: [$$[$0-4], $$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-4], _$[$0]);
        
break;
case 52: case 53: case 54: case 56: case 58: case 60:

        this.$ = $$[$0];
       
break;
case 55:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: !$$[$0-2].value.eq(0) || !$$[$0].value.eq(0) ? bigInt(1) : bigInt(0) };
            } else {
                this.$ = { type: "OP", op: "||", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 57:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: !$$[$0-2].value.eq(0) && !$$[$0].value.eq(0) ? bigInt(1) : bigInt(0) };
            } else {
                this.$ = { type: "OP", op: "&&", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 59:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.or($$[$0].value).and(__MASK__) };
            } else {
                this.$ = { type: "OP", op: "|", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 61:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.xor($$[$0].value).and(__MASK__) };
            } else {
                this.$ = { type: "OP", op: "^", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 63:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.and($$[$0].value).and(__MASK__) };
            } else {
                this.$ = { type: "OP", op: "&", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 65:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.equals($$[$0].value) ? bigInt(1) : bigInt(0) };
            } else {
                this.$ = { type: "OP", op: "==", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 66:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.eq($$[$0].value) ? bigInt(0) : bigInt(1) };
            } else {
                this.$ = { type: "OP", op: "!=", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 67: case 72:

        this.$ = $$[$0]
       
break;
case 68:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.lesserOrEquals($$[$0].value) ? bigInt(1) : bigInt(0) };
            } else {
                this.$ = { type: "OP", op: "<=", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 69:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.greaterOrEquals($$[$0].value) ? bigInt(1) : bigInt(0) };
            } else {
                this.$ = { type: "OP", op: ">=", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 70:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.lesser($$[$0].value) ? bigInt(1) : bigInt(0) };
            } else {
                this.$ = { type: "OP", op: "<", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 71:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.greater($$[$0].value) ? bigInt(1) : bigInt(0) };
            } else {
                this.$ = { type: "OP", op: ">", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 73:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                let v = $$[$0].value.greater(256) ? 256 : $$[$0].value.value;
                this.$ = { type: "NUMBER", value: $$[$0-2].value.shiftLeft(v).and(__MASK__) };
            } else {
                this.$ = { type: "OP", op: "<<", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 74:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                let v = $$[$0].value.greater(256) ? 256 : $$[$0].value.value;
                this.$ = {type: "NUMBER", value: $$[$0-2].value.shiftRight(v).and(__MASK__) };
            } else {
                this.$ = { type: "OP", op: ">>", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 75:

            this.$ = $$[$0];
       
break;
case 76:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: ($$[$0-2].value.plus($$[$0].value)).mod(__P__) };
            } else {
                this.$ = { type: "OP", op: "+", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 77:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: ($$[$0-2].value.plus(__P__).minus($$[$0].value)).mod(__P__) };
            } else {
                this.$ = { type: "OP", op: "-", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 79:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: ($$[$0-2].value.times($$[$0].value)).mod(__P__) };
            } else {
                this.$ = { type: "OP", op: "*", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 80:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: ($$[$0-2].value.times($$[$0].value.modInv(__P__))).mod(__P__) };
            } else {
                this.$ = { type: "OP", op: "/", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 81:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: ($$[$0-2].value.divide($$[$0].value)) };
            } else {
                this.$ = { type: "OP", op: "\\", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 82:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.mod($$[$0].value) };
            } else {
                this.$ = { type: "OP", op: "%", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 84:

            if (($$[$0-2].type == "NUMBER") && ($$[$0].type == "NUMBER")) {
                this.$ = { type: "NUMBER", value: $$[$0-2].value.modPow($$[$0].value, __P__) };
            } else {
                this.$ = { type: "OP", op: "**", values: [$$[$0-2], $$[$0]] };
            }
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 86:

            this.$ = { type: "OP", op: "PLUSPLUSLEFT", values: [$$[$0]] };
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 87:

            this.$ = { type: "OP", op: "MINUSMINUSLEFT", values: [$$[$0]] };
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 88:

            this.$ = $$[$0];
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 89:

            if ($$[$0].type == "NUMBER") {
                this.$ = { type: "NUMBER", value: __P__.minus($$[$0].value).mod(__P__) };
            } else {
                this.$ = { type: "OP", op: "UMINUS", values: [$$[$0]] };
            }
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 90:

            if ($$[$0].type == "NUMBER") {
                this.$ = { type: "NUMBER", value: $$[$0].value.eq(0) ? bigInt(1) : bigInt(0) };
            } else {
                this.$ = { type: "OP", op: "!", values: [$$[$0]] };
            }
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 91:

            if ($$[$0].type == "NUMBER") {
                this.$ = { type: "NUMBER", value: $$[$0].value.xor(__MASK__) };
            } else {
                this.$ = { type: "OP", op: "~", values: [$$[$0]] };
            }
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 93:

            this.$ = {type: "OP", op: "PLUSPLUSRIGHT", values: [$$[$0-1]] };
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 94:

            this.$ = {type: "OP", op: "MINUSMINUSRIGHT", values: [$$[$0-1]] };
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 97: case 102: case 103:

            this.$ = $$[$0]
        
break;
case 98:

            this.$ = {type: "NUMBER", value: bigInt($$[$0]).mod(__P__) }
            setLines(this.$, _$[$0]);
        
break;
case 99:

            this.$ = {type: "NUMBER", value: bigInt($$[$0].substr(2).toUpperCase(), 16).mod(__P__) }
            setLines(this.$, _$[$0]);
        
break;
case 100:

            this.$ = $$[$0-1];
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 101:

            this.$ = {type: "PIN", component: $$[$0-2], pin: $$[$0] };
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 104:

            this.$ = {type: "DECLARE", kind: "T_FIELD", declareType: "VARIABLE", name: $$[$0]}
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 105:

            this.$ = {type: "DECLARE", kind: "T_INT", declareType: "VARIABLE", name: $$[$0]}
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 106:

            this.$ = {type: "DECLARE", kind: "T_FIELD", declareType: "SIGNAL", name: $$[$0]}
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 107:

            this.$ = {type: "DECLARE", kind: "T_FIELD", declareType: "SIGNALIN", name: $$[$0]};
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 108:

            this.$ = {type: "DECLARE", kind: "T_FIELD", declareType: "SIGNALIN", private: true, name: $$[$0]};
            setLines(this.$, _$[$0-3], _$[$0]);
        
break;
case 109:

            this.$ = {type: "DECLARE", kind: "T_FIELD", declareType: "SIGNALOUT", name: $$[$0]};
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 110:

            this.$ = {type: "DECLARE", kind: "T_COMPONENT", declareType: "COMPONENT", name: $$[$0]}
            setLines(this.$, _$[$0-1], _$[$0]);
        
break;
case 111:

            for (let i=0; i< $$[$0].values.length; i++) {
                $$[$0-1].selectors.push($$[$0].values[i]);
            }
            setLines($$[$0-1], _$[$0-1], _$[$0]);
        
break;
case 112:

            this.$ = {type: "VARIABLE", name: $$[$0] , selectors: []};
            setLines(this.$, _$[$0]);
        
break;
case 113:

            this.$ = {type: "FIELDCAST", expression: $$[$0-1]}
            setLines(this.$, _$[$0-3], _$[$0]);
        
break;
case 114:

            this.$ = {type: "INTCAST", expression: $$[$0-1]}
            setLines(this.$, _$[$0-3], _$[$0]);
        
break;
case 115:

            this.$ = {type: "LOG", expression: $$[$0-1]}
            setLines(this.$, _$[$0-3], _$[$0]);
        
break;
case 116:

            this.$ = {type: "FUNCTIONCALL", name: $$[$0-3], params: $$[$0-1].expressions}
            setLines(this.$, _$[$0-3], _$[$0]);
        
break;
case 117:

            this.$ = {type: "FUNCTIONCALL", name: $$[$0-2], params: []}
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 118:

            $$[$0-2].expressions.push($$[$0]);
            setLines(this.$, _$[$0-2], _$[$0]);
        
break;
case 119:

            this.$ = {type: "EXPRESSIONLST", expressions: [$$[$0]]};
            setLines(this.$, _$[$0]);
        
break;
case 121:

                this.$ = { type: "ARRAY", values: $$[$0-1].expressions};
                setLines(this.$, _$[$0-2], _$[$0]);
           
break;
}
},
table: [{3:1,4:2,6:3,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{1:[3]},{5:[1,65],6:66,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($Vr,[2,3]),o($Vs,[2,4]),o($Vs,[2,5]),o($Vs,[2,6]),o($Vs,[2,7]),o($Vs,[2,8]),o($Vs,[2,9]),o($Vs,[2,10]),o($Vs,[2,11]),o($Vs,[2,12]),o($Vs,[2,13]),o($Vs,[2,14]),{19:[1,67]},{19:[1,68]},{20:[1,69]},{20:[1,70]},{20:[1,71]},{6:72,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,26:73,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{4:74,6:3,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{35:[1,75]},o($Vs,[2,33],{29:[1,76]}),{34:[1,77]},o([5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,82,84,92,93,94,95,99,100,104,105,106,110,112,113,115,116],[2,34],{52:$Vt,54:$Vu,57:$Vv}),o($Vw,$Vx,{40:[1,81],41:[1,82],42:[1,83],43:[1,84],44:[1,85],45:[1,86],46:[1,87],47:[1,88],48:[1,89],49:[1,90],50:[1,91],51:[1,92],53:[1,93],92:$Vy,93:$Vz}),o($VA,[2,52],{56:[1,96]}),o($VB,[2,103],{111:98,102:[1,97],115:$Vq}),o([5,18,19,20,22,23,24,25,27,28,29,30,31,32,33,35,36,37,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,79,81,82,84,85,87,88,89,90,92,93,94,95,99,100,104,105,106,110,112,113,115,116],[2,102]),o($VC,[2,53]),o($VC,[2,54],{61:[1,99]}),o([5,18,19,22,23,24,25,27,28,29,30,31,32,33,35,36,37,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,56,57,58,61,63,65,67,69,71,73,74,76,77,78,79,81,82,84,85,87,88,89,90,92,93,94,95,99,100,102,104,105,106,110,112,113,115,116],$VD,{20:[1,100]}),{19:$VE,101:101},{19:$VE,20:[1,104],101:103},{19:$VE,101:105,107:[1,106],108:[1,107],109:[1,108]},{19:$VE,101:109},o($VC,[2,120]),o($VF,[2,56],{63:$VG}),{19:$V1,20:$V2,26:112,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,114:111,115:$Vq},o($VH,[2,58],{65:$VI}),o($VJ,[2,60],{67:$VK}),o($VL,[2,62],{69:$VM}),o($VN,[2,64],{71:$VO,73:$VP}),o($VQ,[2,67],{74:$VR,76:$VS,77:$VT,78:$VU}),o($VV,[2,72],{79:$VW,81:$VX}),o($VY,[2,75],{82:$VZ,84:$V_}),o($V$,[2,78],{85:$V01,87:$V11,88:$V21,89:$V31}),o($V41,[2,83],{90:$V51}),o($V61,[2,85]),{19:$VE,39:131,101:29,103:30,104:$Vk,105:$V71,106:$Vm,110:$Vn},{19:$VE,39:133,101:29,103:30,104:$Vk,105:$V71,106:$Vm,110:$Vn},{19:$V1,20:$V2,39:135,82:$Vc,84:$Vd,91:134,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,84:$Vd,91:136,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,84:$Vd,91:137,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,84:$Vd,91:138,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},o($V61,[2,92]),o($V61,[2,95]),o($V61,[2,96]),{20:[1,139]},{20:[1,140]},o($V61,[2,98]),o($V61,[2,99]),{19:$V1,20:$V2,26:141,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{1:[2,1]},o($Vr,[2,2]),{20:[1,142]},{20:[1,143]},{19:$V1,20:$V2,26:144,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,26:145,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,26:146,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{30:[1,147]},o($Vs,[2,27],{29:[1,148]}),{6:66,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,36:[1,149],37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{4:150,6:3,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($Vs,[2,32]),o($Vs,[2,29],{29:[1,151]}),{19:$VE,39:152,101:29,103:30,104:$Vk,105:$V71,106:$Vm,110:$Vn},{19:$VE,39:153,101:29,103:30,104:$Vk,105:$V71,106:$Vm,110:$Vn},{19:$V1,20:$V2,38:154,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:155,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:156,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:157,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:158,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:159,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:160,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:161,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:162,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:163,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:164,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:165,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:166,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,38:167,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($V61,[2,93]),o($V61,[2,94]),{19:$V1,20:$V2,38:168,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$VE,101:169},o($V81,[2,111]),{19:$V1,20:$V2,39:135,62:170,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,22:[1,172],26:112,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,114:171,115:$Vq},o($VB,[2,104],{111:98,115:$Vq}),o($V81,$VD),o($VB,[2,105],{111:98,115:$Vq}),{19:$V1,20:$V2,26:173,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($VB,[2,106],{111:98,115:$Vq}),{19:$VE,101:174},{107:[1,175]},{19:$VE,101:176},o($VB,[2,110],{111:98,115:$Vq}),{19:$V1,20:$V2,39:135,64:177,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{24:$V91,116:[1,178]},o($Va1,[2,119]),{19:$V1,20:$V2,39:135,66:180,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,68:181,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,70:182,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,72:183,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,72:184,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,75:185,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,75:186,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,75:187,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,75:188,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,80:189,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,80:190,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,83:191,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,83:192,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,84:$Vd,86:193,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,84:$Vd,86:194,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,84:$Vd,86:195,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,84:$Vd,86:196,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},{19:$V1,20:$V2,39:135,82:$Vc,84:$Vd,91:197,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,112:$Vo,113:$Vp},o($V61,[2,86]),{19:$VE,101:103},o($V61,[2,87]),o($V61,[2,88]),o($Vw,$Vx,{92:$Vy,93:$Vz}),o($V61,[2,89]),o($V61,[2,90]),o($V61,[2,91]),{19:$V1,20:$V2,26:198,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,26:199,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{22:[1,200]},{19:$Vb1,21:201,22:[1,202]},{19:$Vb1,21:204,22:[1,205]},{22:[1,206]},{29:[1,207]},{22:[1,208]},{20:[1,209]},o($Vs,[2,26]),o($Vs,[2,30]),{6:66,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,36:[1,210],37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($Vs,[2,28]),o($VA,[2,47]),o($VA,[2,49]),{52:$Vt,54:$Vu,57:$Vv,58:[1,211]},o($Vc1,[2,35],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,36],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,37],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,38],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,39],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,40],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,41],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,42],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,43],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,44],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,45],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,46],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,48],{52:$Vt,54:$Vu,57:$Vv}),o($Vc1,[2,50],{52:$Vt,54:$Vu,57:$Vv}),o($VB,[2,101],{111:98,115:$Vq}),o($VF,[2,55],{63:$VG}),{22:[1,212],24:$V91},o($V61,[2,117]),{22:[1,213]},o($VB,[2,107],{111:98,115:$Vq}),{19:$VE,101:214},o($VB,[2,109],{111:98,115:$Vq}),o($VH,[2,57],{65:$VI}),o($V81,[2,121]),{19:$V1,20:$V2,26:215,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($VJ,[2,59],{67:$VK}),o($VL,[2,61],{69:$VM}),o($VN,[2,63],{71:$VO,73:$VP}),o($VQ,[2,65],{74:$VR,76:$VS,77:$VT,78:$VU}),o($VQ,[2,66],{74:$VR,76:$VS,77:$VT,78:$VU}),o($VV,[2,68],{79:$VW,81:$VX}),o($VV,[2,69],{79:$VW,81:$VX}),o($VV,[2,70],{79:$VW,81:$VX}),o($VV,[2,71],{79:$VW,81:$VX}),o($VY,[2,73],{82:$VZ,84:$V_}),o($VY,[2,74],{82:$VZ,84:$V_}),o($V$,[2,76],{85:$V01,87:$V11,88:$V21,89:$V31}),o($V$,[2,77],{85:$V01,87:$V11,88:$V21,89:$V31}),o($V41,[2,79],{90:$V51}),o($V41,[2,80],{90:$V51}),o($V41,[2,81],{90:$V51}),o($V41,[2,82],{90:$V51}),o($V61,[2,84]),{22:[1,216]},{22:[1,217]},o($V61,[2,100]),{22:[1,218],24:$Vd1},{14:220,35:$Va},o($Ve1,[2,20]),{22:[1,221],24:$Vd1},{14:222,35:$Va},{6:223,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,26:224,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{6:225,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,26:226,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($Vs,[2,31]),{19:$V1,20:$V2,38:227,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($V61,[2,116]),o($V61,[2,114]),o($VB,[2,108],{111:98,115:$Vq}),o($Va1,[2,118]),o($V61,[2,113]),o($V61,[2,115]),{14:228,35:$Va},{19:[1,229]},o($Vs,[2,16]),{14:230,35:$Va},o($Vs,[2,18]),o($Vr,[2,22],{27:[1,231]}),{29:[1,232]},o($Vs,[2,24]),{22:[1,233]},o($Vc1,[2,51],{52:$Vt,54:$Vu,57:$Vv}),o($Vs,[2,15]),o($Ve1,[2,19]),o($Vs,[2,17]),{6:234,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},{19:$V1,20:$V2,26:235,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($Vs,[2,25]),o($Vs,[2,21]),{22:[1,236]},{6:237,7:4,8:5,9:6,10:7,11:8,12:9,13:10,14:11,15:12,16:13,17:14,18:$V0,19:$V1,20:$V2,23:$V3,25:$V4,26:24,28:$V5,30:$V6,31:$V7,32:$V8,33:$V9,35:$Va,37:$Vb,38:26,39:27,55:28,59:31,60:32,62:39,64:41,66:42,68:43,70:44,72:45,75:46,80:47,82:$Vc,83:48,84:$Vd,86:49,91:50,92:$Ve,93:$Vf,94:$Vg,95:$Vh,96:57,97:58,98:59,99:$Vi,100:$Vj,101:29,103:30,104:$Vk,105:$Vl,106:$Vm,110:$Vn,111:38,112:$Vo,113:$Vp,115:$Vq},o($Vs,[2,23])],
defaultActions: {65:[2,1]},
parseError: function parseError (str, hash) {
    if (hash.recoverable) {
        this.trace(str);
    } else {
        var error = new Error(str);
        error.hash = hash;
        throw error;
    }
},
parse: function parse(input) {
    var self = this, stack = [0], tstack = [], vstack = [null], lstack = [], table = this.table, yytext = '', yylineno = 0, yyleng = 0, recovering = 0, TERROR = 2, EOF = 1;
    var args = lstack.slice.call(arguments, 1);
    var lexer = Object.create(this.lexer);
    var sharedState = { yy: {} };
    for (var k in this.yy) {
        if (Object.prototype.hasOwnProperty.call(this.yy, k)) {
            sharedState.yy[k] = this.yy[k];
        }
    }
    lexer.setInput(input, sharedState.yy);
    sharedState.yy.lexer = lexer;
    sharedState.yy.parser = this;
    if (typeof lexer.yylloc == 'undefined') {
        lexer.yylloc = {};
    }
    var yyloc = lexer.yylloc;
    lstack.push(yyloc);
    var ranges = lexer.options && lexer.options.ranges;
    if (typeof sharedState.yy.parseError === 'function') {
        this.parseError = sharedState.yy.parseError;
    } else {
        this.parseError = Object.getPrototypeOf(this).parseError;
    }
    function popStack(n) {
        stack.length = stack.length - 2 * n;
        vstack.length = vstack.length - n;
        lstack.length = lstack.length - n;
    }
    _token_stack:
        var lex = function () {
            var token;
            token = lexer.lex() || EOF;
            if (typeof token !== 'number') {
                token = self.symbols_[token] || token;
            }
            return token;
        };
    var symbol, preErrorSymbol, state, action, a, r, yyval = {}, p, len, newState, expected;
    while (true) {
        state = stack[stack.length - 1];
        if (this.defaultActions[state]) {
            action = this.defaultActions[state];
        } else {
            if (symbol === null || typeof symbol == 'undefined') {
                symbol = lex();
            }
            action = table[state] && table[state][symbol];
        }
                    if (typeof action === 'undefined' || !action.length || !action[0]) {
                var errStr = '';
                expected = [];
                for (p in table[state]) {
                    if (this.terminals_[p] && p > TERROR) {
                        expected.push('\'' + this.terminals_[p] + '\'');
                    }
                }
                if (lexer.showPosition) {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ':\n' + lexer.showPosition() + '\nExpecting ' + expected.join(', ') + ', got \'' + (this.terminals_[symbol] || symbol) + '\'';
                } else {
                    errStr = 'Parse error on line ' + (yylineno + 1) + ': Unexpected ' + (symbol == EOF ? 'end of input' : '\'' + (this.terminals_[symbol] || symbol) + '\'');
                }
                this.parseError(errStr, {
                    text: lexer.match,
                    token: this.terminals_[symbol] || symbol,
                    line: lexer.yylineno,
                    loc: yyloc,
                    expected: expected
                });
            }
        if (action[0] instanceof Array && action.length > 1) {
            throw new Error('Parse Error: multiple actions possible at state: ' + state + ', token: ' + symbol);
        }
        switch (action[0]) {
        case 1:
            stack.push(symbol);
            vstack.push(lexer.yytext);
            lstack.push(lexer.yylloc);
            stack.push(action[1]);
            symbol = null;
            if (!preErrorSymbol) {
                yyleng = lexer.yyleng;
                yytext = lexer.yytext;
                yylineno = lexer.yylineno;
                yyloc = lexer.yylloc;
                if (recovering > 0) {
                    recovering--;
                }
            } else {
                symbol = preErrorSymbol;
                preErrorSymbol = null;
            }
            break;
        case 2:
            len = this.productions_[action[1]][1];
            yyval.$ = vstack[vstack.length - len];
            yyval._$ = {
                first_line: lstack[lstack.length - (len || 1)].first_line,
                last_line: lstack[lstack.length - 1].last_line,
                first_column: lstack[lstack.length - (len || 1)].first_column,
                last_column: lstack[lstack.length - 1].last_column
            };
            if (ranges) {
                yyval._$.range = [
                    lstack[lstack.length - (len || 1)].range[0],
                    lstack[lstack.length - 1].range[1]
                ];
            }
            r = this.performAction.apply(yyval, [
                yytext,
                yyleng,
                yylineno,
                sharedState.yy,
                action[1],
                vstack,
                lstack
            ].concat(args));
            if (typeof r !== 'undefined') {
                return r;
            }
            if (len) {
                stack = stack.slice(0, -1 * len * 2);
                vstack = vstack.slice(0, -1 * len);
                lstack = lstack.slice(0, -1 * len);
            }
            stack.push(this.productions_[action[1]][0]);
            vstack.push(yyval.$);
            lstack.push(yyval._$);
            newState = table[stack[stack.length - 2]][stack[stack.length - 1]];
            stack.push(newState);
            break;
        case 3:
            return true;
        }
    }
    return true;
}};

const bigInt = require('big-integer');
const util = require('util');
const __P__ = new bigInt("21888242871839275222246405745257275088548364400416034343698204186575808495617");
const __MASK__ = new bigInt(2).pow(253).minus(1);

function setLines(dst, first, last) {
    last = last || first;
    dst.first_line = first.first_line;
    dst.first_column = first.first_column;
    dst.last_line = last.last_line;
    dst.last_column = last.last_column;
}

/* generated by jison-lex 0.3.4 */
var lexer = (function(){
var lexer = ({

EOF:1,

parseError:function parseError(str, hash) {
        if (this.yy.parser) {
            this.yy.parser.parseError(str, hash);
        } else {
            throw new Error(str);
        }
    },

// resets the lexer, sets new input
setInput:function (input, yy) {
        this.yy = yy || this.yy || {};
        this._input = input;
        this._more = this._backtrack = this.done = false;
        this.yylineno = this.yyleng = 0;
        this.yytext = this.matched = this.match = '';
        this.conditionStack = ['INITIAL'];
        this.yylloc = {
            first_line: 1,
            first_column: 0,
            last_line: 1,
            last_column: 0
        };
        if (this.options.ranges) {
            this.yylloc.range = [0,0];
        }
        this.offset = 0;
        return this;
    },

// consumes and returns one char from the input
input:function () {
        var ch = this._input[0];
        this.yytext += ch;
        this.yyleng++;
        this.offset++;
        this.match += ch;
        this.matched += ch;
        var lines = ch.match(/(?:\r\n?|\n).*/g);
        if (lines) {
            this.yylineno++;
            this.yylloc.last_line++;
        } else {
            this.yylloc.last_column++;
        }
        if (this.options.ranges) {
            this.yylloc.range[1]++;
        }

        this._input = this._input.slice(1);
        return ch;
    },

// unshifts one char (or a string) into the input
unput:function (ch) {
        var len = ch.length;
        var lines = ch.split(/(?:\r\n?|\n)/g);

        this._input = ch + this._input;
        this.yytext = this.yytext.substr(0, this.yytext.length - len);
        //this.yyleng -= len;
        this.offset -= len;
        var oldLines = this.match.split(/(?:\r\n?|\n)/g);
        this.match = this.match.substr(0, this.match.length - 1);
        this.matched = this.matched.substr(0, this.matched.length - 1);

        if (lines.length - 1) {
            this.yylineno -= lines.length - 1;
        }
        var r = this.yylloc.range;

        this.yylloc = {
            first_line: this.yylloc.first_line,
            last_line: this.yylineno + 1,
            first_column: this.yylloc.first_column,
            last_column: lines ?
                (lines.length === oldLines.length ? this.yylloc.first_column : 0)
                 + oldLines[oldLines.length - lines.length].length - lines[0].length :
              this.yylloc.first_column - len
        };

        if (this.options.ranges) {
            this.yylloc.range = [r[0], r[0] + this.yyleng - len];
        }
        this.yyleng = this.yytext.length;
        return this;
    },

// When called from action, caches matched text and appends it on next action
more:function () {
        this._more = true;
        return this;
    },

// When called from action, signals the lexer that this rule fails to match the input, so the next matching rule (regex) should be tested instead.
reject:function () {
        if (this.options.backtrack_lexer) {
            this._backtrack = true;
        } else {
            return this.parseError('Lexical error on line ' + (this.yylineno + 1) + '. You can only invoke reject() in the lexer when the lexer is of the backtracking persuasion (options.backtrack_lexer = true).\n' + this.showPosition(), {
                text: "",
                token: null,
                line: this.yylineno
            });

        }
        return this;
    },

// retain first n characters of the match
less:function (n) {
        this.unput(this.match.slice(n));
    },

// displays already matched input, i.e. for error messages
pastInput:function () {
        var past = this.matched.substr(0, this.matched.length - this.match.length);
        return (past.length > 20 ? '...':'') + past.substr(-20).replace(/\n/g, "");
    },

// displays upcoming input, i.e. for error messages
upcomingInput:function () {
        var next = this.match;
        if (next.length < 20) {
            next += this._input.substr(0, 20-next.length);
        }
        return (next.substr(0,20) + (next.length > 20 ? '...' : '')).replace(/\n/g, "");
    },

// displays the character position where the lexing error occurred, i.e. for error messages
showPosition:function () {
        var pre = this.pastInput();
        var c = new Array(pre.length + 1).join("-");
        return pre + this.upcomingInput() + "\n" + c + "^";
    },

// test the lexed token: return FALSE when not a match, otherwise return token
test_match:function(match, indexed_rule) {
        var token,
            lines,
            backup;

        if (this.options.backtrack_lexer) {
            // save context
            backup = {
                yylineno: this.yylineno,
                yylloc: {
                    first_line: this.yylloc.first_line,
                    last_line: this.last_line,
                    first_column: this.yylloc.first_column,
                    last_column: this.yylloc.last_column
                },
                yytext: this.yytext,
                match: this.match,
                matches: this.matches,
                matched: this.matched,
                yyleng: this.yyleng,
                offset: this.offset,
                _more: this._more,
                _input: this._input,
                yy: this.yy,
                conditionStack: this.conditionStack.slice(0),
                done: this.done
            };
            if (this.options.ranges) {
                backup.yylloc.range = this.yylloc.range.slice(0);
            }
        }

        lines = match[0].match(/(?:\r\n?|\n).*/g);
        if (lines) {
            this.yylineno += lines.length;
        }
        this.yylloc = {
            first_line: this.yylloc.last_line,
            last_line: this.yylineno + 1,
            first_column: this.yylloc.last_column,
            last_column: lines ?
                         lines[lines.length - 1].length - lines[lines.length - 1].match(/\r?\n?/)[0].length :
                         this.yylloc.last_column + match[0].length
        };
        this.yytext += match[0];
        this.match += match[0];
        this.matches = match;
        this.yyleng = this.yytext.length;
        if (this.options.ranges) {
            this.yylloc.range = [this.offset, this.offset += this.yyleng];
        }
        this._more = false;
        this._backtrack = false;
        this._input = this._input.slice(match[0].length);
        this.matched += match[0];
        token = this.performAction.call(this, this.yy, this, indexed_rule, this.conditionStack[this.conditionStack.length - 1]);
        if (this.done && this._input) {
            this.done = false;
        }
        if (token) {
            return token;
        } else if (this._backtrack) {
            // recover context
            for (var k in backup) {
                this[k] = backup[k];
            }
            return false; // rule action called reject() implying the next rule should be tested instead.
        }
        return false;
    },

// return next match in input
next:function () {
        if (this.done) {
            return this.EOF;
        }
        if (!this._input) {
            this.done = true;
        }

        var token,
            match,
            tempMatch,
            index;
        if (!this._more) {
            this.yytext = '';
            this.match = '';
        }
        var rules = this._currentRules();
        for (var i = 0; i < rules.length; i++) {
            tempMatch = this._input.match(this.rules[rules[i]]);
            if (tempMatch && (!match || tempMatch[0].length > match[0].length)) {
                match = tempMatch;
                index = i;
                if (this.options.backtrack_lexer) {
                    token = this.test_match(tempMatch, rules[i]);
                    if (token !== false) {
                        return token;
                    } else if (this._backtrack) {
                        match = false;
                        continue; // rule action called reject() implying a rule MISmatch.
                    } else {
                        // else: this is a lexer rule which consumes input without producing a token (e.g. whitespace)
                        return false;
                    }
                } else if (!this.options.flex) {
                    break;
                }
            }
        }
        if (match) {
            token = this.test_match(match, rules[index]);
            if (token !== false) {
                return token;
            }
            // else: this is a lexer rule which consumes input without producing a token (e.g. whitespace)
            return false;
        }
        if (this._input === "") {
            return this.EOF;
        } else {
            return this.parseError('Lexical error on line ' + (this.yylineno + 1) + '. Unrecognized text.\n' + this.showPosition(), {
                text: "",
                token: null,
                line: this.yylineno
            });
        }
    },

// return next match that has a token
lex:function lex () {
        var r = this.next();
        if (r) {
            return r;
        } else {
            return this.lex();
        }
    },

// activates a new lexer condition state (pushes the new lexer condition state onto the condition stack)
begin:function begin (condition) {
        this.conditionStack.push(condition);
    },

// pop the previously active lexer condition state off the condition stack
popState:function popState () {
        var n = this.conditionStack.length - 1;
        if (n > 0) {
            return this.conditionStack.pop();
        } else {
            return this.conditionStack[0];
        }
    },

// produce the lexer rule set which is active for the currently active lexer condition state
_currentRules:function _currentRules () {
        if (this.conditionStack.length && this.conditionStack[this.conditionStack.length - 1]) {
            return this.conditions[this.conditionStack[this.conditionStack.length - 1]].rules;
        } else {
            return this.conditions["INITIAL"].rules;
        }
    },

// return the currently active lexer condition state; when an index argument is provided it produces the N-th previous condition state, if available
topState:function topState (n) {
        n = this.conditionStack.length - 1 - Math.abs(n || 0);
        if (n >= 0) {
            return this.conditionStack[n];
        } else {
            return "INITIAL";
        }
    },

// alias for begin(condition)
pushState:function pushState (condition) {
        this.begin(condition);
    },

// return the number of states currently on the stack
stateStackSize:function stateStackSize() {
        return this.conditionStack.length;
    },
options: {},
performAction: function anonymous(yy,yy_,$avoiding_name_collisions,YY_START) {
var YYSTATE=YY_START;
switch($avoiding_name_collisions) {
case 0: /* skip whitespace */ 
break;
case 1: /* console.log("MULTILINE COMMENT: "+yy_.yytext); */  
break;
case 2: /* console.log("SINGLE LINE COMMENT: "+yy_.yytext); */ 
break;
case 3: return 104; 
break;
case 4: return 106; 
break;
case 5: return 108; 
break;
case 6: return 107; 
break;
case 7: return 109; 
break;
case 8: return 'linearCombination'; 
break;
case 9: return 112; 
break;
case 10: return 105; 
break;
case 11: return 113; 
break;
case 12: return 110; 
break;
case 13: return 37; 
break;
case 14: return 23; 
break;
case 15: return 18; 
break;
case 16: return 25; 
break;
case 17: return 27; 
break;
case 18: return 28; 
break;
case 19: return 30; 
break;
case 20: return 31; 
break;
case 21: return 32; 
break;
case 22: return 33; 
break;
case 23: return 100; 
break;
case 24: return 99; 
break;
case 25: return 19; 
break;
case 26: yy_.yytext = yy_.yytext.slice(1,-1); return 34; 
break;
case 27: return 52; 
break;
case 28: return 51; 
break;
case 29: return 54; 
break;
case 30: return 53; 
break;
case 31: return 56; 
break;
case 32: return 47; 
break;
case 33: return 46; 
break;
case 34: return 63; 
break;
case 35: return 61; 
break;
case 36: return 71; 
break;
case 37: return 74; 
break;
case 38: return 76; 
break;
case 39: return 73; 
break;
case 40: return 81; 
break;
case 41: return 79; 
break;
case 42: return 90; 
break;
case 43: return 92; 
break;
case 44: return 93; 
break;
case 45: return 41; 
break;
case 46: return 42; 
break;
case 47: return 43; 
break;
case 48: return 44; 
break;
case 49: return 45; 
break;
case 50: return 49; 
break;
case 51: return 48; 
break;
case 52: return 50; 
break;
case 53: return 40; 
break;
case 54: return 82; 
break;
case 55: return 84; 
break;
case 56: return 85; 
break;
case 57: return 87; 
break;
case 58: return 88; 
break;
case 59: return 89; 
break;
case 60: return 67; 
break;
case 61: return 69; 
break;
case 62: return 65; 
break;
case 63: return 94; 
break;
case 64: return 77; 
break;
case 65: return 78; 
break;
case 66: return 94; 
break;
case 67: return 57; 
break;
case 68: return 58; 
break;
case 69: return 20; 
break;
case 70: return 22; 
break;
case 71: return 115; 
break;
case 72: return 116; 
break;
case 73: return 35; 
break;
case 74: return 36; 
break;
case 75: return 29; 
break;
case 76: return 24; 
break;
case 77: return 102; 
break;
case 78: return 5; 
break;
case 79: console.log("INVALID: " + yy_.yytext); return 'INVALID'
break;
}
},
rules: [/^(?:\s+)/,/^(?:\/\*([^*]|[\r\n]|(\*+([^*\/]|[\r\n])))*\*+\/)/,/^(?:\/\/.*)/,/^(?:var\b)/,/^(?:signal\b)/,/^(?:private\b)/,/^(?:input\b)/,/^(?:output\b)/,/^(?:linearCombination\b)/,/^(?:field\b)/,/^(?:int\b)/,/^(?:log\b)/,/^(?:component\b)/,/^(?:compute\b)/,/^(?:template\b)/,/^(?:function\b)/,/^(?:if\b)/,/^(?:else\b)/,/^(?:for\b)/,/^(?:while\b)/,/^(?:do\b)/,/^(?:return\b)/,/^(?:include\b)/,/^(?:0x[0-9A-Fa-f]*)/,/^(?:[0-9]+)/,/^(?:[a-zA-Z][a-zA-Z$_0-9]*)/,/^(?:"[^"]+")/,/^(?:==>)/,/^(?:<==)/,/^(?:-->)/,/^(?:<--)/,/^(?:===)/,/^(?:>>=)/,/^(?:<<=)/,/^(?:&&)/,/^(?:\|\|)/,/^(?:==)/,/^(?:<=)/,/^(?:>=)/,/^(?:!=)/,/^(?:>>)/,/^(?:<<)/,/^(?:\*\*)/,/^(?:\+\+)/,/^(?:--)/,/^(?:\+=)/,/^(?:-=)/,/^(?:\*=)/,/^(?:\/=)/,/^(?:%=)/,/^(?:\|=)/,/^(?:&=)/,/^(?:\^=)/,/^(?:=)/,/^(?:\+)/,/^(?:-)/,/^(?:\*)/,/^(?:\/)/,/^(?:\\)/,/^(?:%)/,/^(?:\^)/,/^(?:&)/,/^(?:\|)/,/^(?:!)/,/^(?:<)/,/^(?:>)/,/^(?:!)/,/^(?:\?)/,/^(?::)/,/^(?:\()/,/^(?:\))/,/^(?:\[)/,/^(?:\])/,/^(?:\{)/,/^(?:\})/,/^(?:;)/,/^(?:,)/,/^(?:\.)/,/^(?:$)/,/^(?:.)/],
conditions: {"INITIAL":{"rules":[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79],"inclusive":true}}
});
return lexer;
})();
parser.lexer = lexer;
function Parser () {
  this.yy = {};
}
Parser.prototype = parser;parser.Parser = Parser;
return new Parser;
})();


if (typeof require !== 'undefined' && typeof exports !== 'undefined') {
exports.parser = jaz;
exports.Parser = jaz.Parser;
exports.parse = function () { return jaz.parse.apply(jaz, arguments); };
exports.main = function commonjsMain (args) {
    if (!args[1]) {
        console.log('Usage: '+args[0]+' FILE');
        process.exit(1);
    }
    var source = require('fs').readFileSync(require('path').normalize(args[1]), "utf8");
    return exports.parser.parse(source);
};
if (typeof module !== 'undefined' && require.main === module) {
  exports.main(process.argv.slice(1));
}
}
